### PR TITLE
add troubleshooting-write-conflict

### PR DESCRIPTION
**Pull request by crazycs520**
### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs-cn/pull/3209
- Other reference link(s):

### CODE REVIEW

File: troubleshoot-write-conflicts.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+title: TiDB Write Conflict In Optimistic Transaction Troubleshooting
+summary: Learn about common errors and solutions of write conflict in optimistic transaction.
+category: how-to
+---
+
+# TiDB Write Conflict In Optimistic Transaction Troubleshooting
+
+This article is an introduction to the reason and solutions of write conflict in optimistic transaction.
+
+Before TiDB v3.0.8, optimistic transaction is the defalut transaction model in TiDB. The transaction doesn't check the confliction during the execution, it checks the confliction when committing the transaction. If the write conflict exists when committing the transaction, the transaction will be auto-retry in TiDB internal when TiDB has enabled the auto-retry and the retry count doesn't exceeded the limie. Then, TiDB will return the result of transaction execution to the client. Therefore, the write latency of TiDB cluster will be higher when there are a lot of write conflict exists.
```
Comment (Win-Man) [inline troubleshoot-write-conflicts.md:11]: limie -> limit

File: troubleshoot-write-conflicts.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+After the client send a `COMMIT` request to TiDB, TiDB starts the 2PC process:
+
+1. TiDB chooses one key from all keys in the transaction as the primary key of the transaction.
+2. TiDB sends `prewrite` requests to the TiKV regions which contain the all keys of the transaction.
+3. After all `prewrite` requests return successful result, go to the next step.
+4. TiDB gets the `commit_ts` from PD.
+5. TiDB sends `commit` request to the TiKV region which contain the primary key of the transaction. When TiKV executes `commit` request, TiKV checks the validity of the data and clear the lock that left in `prewrite` request.
+6. After the `commit` request returns successfully, TiDB return success to the client.
+
+The write conflict occurs in the `prewrite` stage. When the transaction found a new version of the key (key.commit_ts > txn.start_ts), that's write conflict.
```
Comment (Win-Man) [inline troubleshoot-write-conflicts.md:26]: key.commit_ts -> data.commit_ts

File: troubleshoot-write-conflicts.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    ![kv-backoff-ops](/media/troubleshooting-write-conflict-kv-backoff-ops.png)
+
+    `txnlock` metric indicates the write-write conflict. `txnLockFast` metric indicates the read-write conflict.
+
+* **Lock Resolve OPS** indicates the count of resolve locks per second which cause by transaction conflict:
+
+    ![lock-resolve-ops](/media/troubleshooting-write-conflict-lock-resolve-ops.png)
+
+    `not_expired` means the TTL of lock was not expired, Then the conflict transaction can't resolve locks until the TTL was expired.
+    `wait_expire` means the transaction needs to wait the lock to expired.
```
Comment (Win-Man) [inline troubleshoot-write-conflicts.md:43]: wait_expire -> wait_expired

File: troubleshoot-write-conflicts.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+# TiDB Write Conflict In Optimistic Transaction Troubleshooting
+
+This article is an introduction to the reason and solutions of write conflict in optimistic transaction.
+
+Before TiDB v3.0.8, optimistic transaction is the defalut transaction model in TiDB. The transaction doesn't check the confliction during the execution, it checks the confliction when committing the transaction. If the write conflict exists when committing the transaction, the transaction will be auto-retry in TiDB internal when TiDB has enabled the auto-retry and the retry count doesn't exceeded the limit. Then, TiDB will return the result of transaction execution to the client. Therefore, the write latency of TiDB cluster will be higher when there are a lot of write conflict exists.
+
+## The reason of write conflict
+
+TiDB use [Percolator](https://www.usenix.org/legacy/event/osdi10/tech/full_papers/Peng.pdf) transaction model as transaction implemention. And `percolator` is generally an implementation of 2PC. Please refer to [optimistic transaction document](/optimistic-transaction.md) for the specific 2PC process.
```
Comment (lilin90) [inline troubleshoot-write-conflicts.md:15]:
```suggestion
TiDB implements its transactions by using the [Percolator](https://www.usenix.org/legacy/event/osdi10/tech/full_papers/Peng.pdf) transaction model. `percolator` is generally an implementation of 2PC. For the detailed 2PC process, see [TiDB Optimistic Transaction Model](/optimistic-transaction.md).
```

File: troubleshoot-write-conflicts.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+After the client send a `COMMIT` request to TiDB, TiDB starts the 2PC process:
```
Comment (lilin90) [inline troubleshoot-write-conflicts.md:17]:
```suggestion
After the client sends a `COMMIT` request to TiDB, TiDB starts the 2PC process:
```

File: troubleshoot-write-conflicts.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+1. TiDB chooses one key from all keys in the transaction as the primary key of the transaction.
+2. TiDB sends `prewrite` requests to the TiKV regions which contain the all keys of the transaction.
```
Comment (lilin90) [inline troubleshoot-write-conflicts.md:20]:
```suggestion
2. TiDB sends the `prewrite` request to all the TiKV Regions involved in this commit. TiKV judges whether all keys can preview successfully.
```

File: troubleshoot-write-conflicts.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+3. After all `prewrite` requests return successful result, go to the next step.
```
Comment (lilin90) [inline troubleshoot-write-conflicts.md:21]:
```suggestion
3. TiDB receives the result that all `prewrite` requests are successful.
```

File: troubleshoot-write-conflicts.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+4. TiDB gets the `commit_ts` from PD.
+5. TiDB sends `commit` request to the TiKV region which contain the primary key of the transaction. When TiKV executes `commit` request, TiKV checks the validity of the data and clear the lock that left in `prewrite` request.
```
Comment (lilin90) [inline troubleshoot-write-conflicts.md:23]:
```suggestion
5. TiDB sends the `commit` request to the TiKV Region that contains the primary key of the transaction. After TiKV receives the `commit` request, it checks the validity of the data and clears the locks left in the `prewrite` stage.
```

File: troubleshoot-write-conflicts.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+6. After the `commit` request returns successfully, TiDB return success to the client.
```
Comment (lilin90) [inline troubleshoot-write-conflicts.md:24]:
```suggestion
6. After the `commit` request returns successfully, TiDB returns success to the client.
```

File: troubleshoot-write-conflicts.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+The write conflict occurs in the `prewrite` stage. When the transaction found a new version of the key (data.commit_ts > txn.start_ts), that's write conflict.
```
Comment (lilin90) [inline troubleshoot-write-conflicts.md:26]:
```suggestion
The write conflict occurs in the `prewrite` stage. When the transaction finds that another transaction is writing the current key (`data.commit_ts` > `txn.start_ts`), a write conflict occurs.
```

File: troubleshoot-write-conflicts.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+## How to detect the write conflict exists in the cluster?
```
Comment (lilin90) [inline troubleshoot-write-conflicts.md:28]:
```suggestion
## Detect write conflicts
```

File: troubleshoot-write-conflicts.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    `wait_expired` means the transaction needs to wait the lock to expired.
+    `expired` means the TTL of lock was expired, Then the conflict transaction can resolve this lock.
+
+* **KV Retry Duration** indicates the duration of re-sends the KV request:
+
+     ![kv-retry-duration](/media/troubleshooting-write-conflict-kv-retry-duration.png)
+
+You can also use `[kv:9007]Write conflict` as the key word to search in TiDB log. The key word also indicates the write conflict exists in the cluster.
+
+## How to resolve the write-conflict?
```
Comment (lilin90) [inline troubleshoot-write-conflicts.md:52]:
```suggestion
## Resolve write conflicts
```

**Comment by lilin90** [comment]
@Win-Man PTAL again~ Thanks!

**Pull request merged**